### PR TITLE
bam_variant_calling_freebayes

### LINKED ISSUE #2215: new subworkflow variantcalling/freebayes
A generic version of the freebayes subworkflow that can be imported to replace steps in [sarek](https://github.com/nf-core/sarek/blob/dev/subworkflows/nf-core/variantcalling/freebayes/main.nf) and [mag](https://github.com/nf-core/mag/blob/dev/subworkflows/local/ancient_dna.nf)

Modules included:
- SAMTOOLS_FAIDX
- FREEBAYES
- BCFTOOLS_SORT
- BCFTOOLS_INDEX

### PR DESCRIPTION
**Pull request by priyanka-surana**
A generic version of the freebayes subworkflow that can be imported to replace steps in [sarek](https://github.com/nf-core/sarek/blob/dev/subworkflows/nf-core/variantcalling/freebayes/main.nf) and [mag](https://github.com/nf-core/mag/blob/dev/subworkflows/local/ancient_dna.nf)

Modules included:
- SAMTOOLS_FAIDX
- FREEBAYES
- BCFTOOLS_SORT
- BCFTOOLS_INDEX

## PR checklist

Closes #2215

- [ x] This comment contains a description of changes (with reason).
- [ x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ x] Remove all TODO statements.
- [ x] Emit the `versions.yml` file.
- [ x] Follow the naming conventions.
- [ x] Follow the parameters requirements.
- [ x] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (matthdsm) [inline subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:21]: Shouldn't we take `.fai` as an input and only generate it of it's not available? Seems like unnecessary overhead to me

File: subworkflows/nf-core/bam_variant_calling_freebayes/meta.yml
Comment (matthdsm) [inline subworkflows/nf-core/bam_variant_calling_freebayes/meta.yml:77]: If I'm not mistaken, the `csi` index is for larger files, not the `tbi`

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (jfy133) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:10]: Agreed

**Comment by matthdsm** [comment]
Could you also add a small readme describing the workflow steps in a diagram?
https://mermaid-js.github.io/mermaid/#/
https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

File: subworkflows/nf-core/bam_variant_calling_freebayes/meta.yml
Comment (Emiller88) [inline subworkflows/nf-core/bam_variant_calling_freebayes/meta.yml:80]:
```suggestion
  - "@priyanka-surana"
  - "@FriederikeHanssen"
```
@FriederikeHanssen maybe? I think we should swap authors to "maintainers".

File: tests/subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (Emiller88) [inline tests/subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:21]: Can we add another test showing use with all the expected inputs?

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (Emiller88) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:21]: Agreed. 
It could be merged with the fasta FYI, `[meta, fasta, fai]`. But kinda of pointless unless you're going to merge all of them on `ch_input`

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (jfy133) [inline subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:9]: What is bam/cram, bai/crai there twice?

**Review by Emiller88 — CHANGES_REQUESTED** [review]

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (maxulysse) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:9]: For somatic variant calling, it's in the tests, we discussed about that already, we merged it months ago

Comment (jfy133) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:9]: Maybe label accordingly :)

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (priyanka-surana) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:21]: Since I can only implement this or the below suggestion, I went with `ch_genome  // channel: [ val(meta), fasta, fai ]`

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (priyanka-surana) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:10]: Since I can only implement this or the previous suggestion, I went with `ch_genome  // channel: [ val(meta), fasta, fai ]`

File: subworkflows/nf-core/bam_variant_calling_freebayes/meta.yml
Comment (priyanka-surana) [reply subworkflows/nf-core/bam_variant_calling_freebayes/meta.yml:77]: I borrowed from [`bcftools_index`](https://github.com/nf-core/modules/blob/master/modules/nf-core/bcftools/index/meta.yml#L39), but you are right.
> CSI (coordinate-sorted index) is created by default. The CSI format supports indexing of chromosomes up to length 2^31. TBI (tabix index) index files, which support chromosome lengths up to 2^29. When loading an index file, bcftools will try the CSI first and then the TBI.
I have updated but I `bcftools_index` needs to be updated too.

File: tests/subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (priyanka-surana) [reply tests/subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:21]: I don't have `samples`, `populations` or `cnv` examples to create a test with all inputs. None in [`freebayes`](https://github.com/nf-core/modules/blob/master/tests/modules/nf-core/freebayes/main.nf) either.

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (priyanka-surana) [reply subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:9]: Added comments in `meta.yml` to clarify.

**Comment by priyanka-surana** [comment]
> Could you also add a small readme describing the workflow steps in a diagram?
> https://mermaid-js.github.io/mermaid/#/
> https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

Where should I add this information?

**Comment by matthdsm** [comment]
A readme.md file next to `main.nf` seems allright 🤷🏻‍♂️ 
@Emiller88 , opinion?

**Comment by Emiller88** [comment]
Yeah, I think a `README.md` in the dir should be the move. It should get rendered by GitHub then!

**Comment by priyanka-surana** [comment]
> Yeah, I think a `README.md` in the dir should be the move. It should get rendered by GitHub then!

Should I wait till the tooling update is made or should I manually add it for this subworkflow?

**Comment by Emiller88** [comment]
> Should I wait till the tooling update is made or should I manually add it for this subworkflow?

Manually add it!

**Comment by priyanka-surana** [comment]
@Emiller88 @matthdsm I added a `readme` file with the `mermaid` flowchart. I am not at all familiar with mermaid so please make changes. Unless some of the `meta.yml` information is moving to `readme.md`, maybe call it a `mermaid_flowchart.md` or `mermaid.md`, much more clear.

**Comment by matthdsm** [comment]
Hi @priyanka-surana 
If you put the flowchart code inside a code block with the `mermaid` designator, it will automatically be rendered.
so use
```
\```mermaid
\ flowchart TD
\ ...
\```
```
(ignore the backslashes)

File: subworkflows/nf-core/bam_variant_calling_freebayes/main.nf
Comment (matthdsm) [inline subworkflows/nf-core/bam_variant_calling_freebayes/main.nf:9]:
```suggestion
    ch_fasta_fai       // channel: [ val(meta), fasta, fai ]
```
Let's bring this in line with other modules

**Review by matthdsm — APPROVED** [review]

**Comment by matthdsm** [comment]
Hi there!

We've noticed there hasn't been much activity here. Are you still planning on working on this?
If not, you can ignore this message and we'll close your PR in about 2 weeks

Cheers
the nf-core maintainers

**Comment by priyanka-surana** [comment]
The merging is blocked. I am not sure what additional changes are needed to merge this. I think that is why it got inactive.

**Comment by matthdsm** [comment]
Thanks for the reply @priyanka-surana, I've removed the stale review.
Feel free to merge once the tests pass!

Matthias

**Comment by priyanka-surana** [comment]
I think the md5sums need updating. Will commit the changes later and merge.

**Pull request merged**
